Event order page

Kakadu opened the pull request:
Timeline template has now a link to the dedicated page for events reordering.
Events without a date can be moved up and down, dated events are always in the
same order. After applying changes by user new order is sent to the server
where new order is being validated and commited.

Page for updating a person is slightly changed. Event positing in the list
(a.k.a. weight) is exposed to the form. Form hadling code uses this weight to
reconstitute order of the events and to distinguish old events from new ones.
Before that all new events were automatically sorted. A warning about changed
event order is removed.

Also type alias for cdate was removed.

a2line commented:
Petit retour :
A2@D ~/master
$ curl -O
https://patch-diff.githubusercontent.com/raw/geneanet/geneweb/pull/349.patch
  % Total    % Received % Xferd  Average Speed   Time    Time     Time
 Current
                                 Dload  Upload   Total   Spent    Left
 Speed
100 38183    0 38183    0     0  25138      0 --:--:--  0:00:01 --:--:--
55579

A2@D ~/master
$ git apply 349.patch
349.patch:1061: new blank line at EOF.
+
warning: 1 line adds whitespace errors.

puis make/make distrib puis import d'un .gw 6.08 sans événement auquel
j'ajoute un événement sans date avec le template par défaut, je clique sur
le nouveau lien “Edit person events order” au pied de la chronologie qui
m’ouvre

http://localhost:2317/a2?m=MOD_EVENT_ORDER;i=11 avec en entête « Requête
incorrecte ». Pareil quelque soit la fiche ou l'événement ajouté. Le log du
serv gwd.exe donne chaque fois trois fois :
2015-08-27 16:11:51 (88) ?
  From: 127.0.0.1
"read" failed: An established connection was aborted by the software in
your host machine.

Je me suis planté quelque part ?

– A2

2015-08-27 1:39 GMT+02:00 Kakadu notifications@github.com:

> Timeline template has now a link to the dedicated page for events
> reordering.
> Events without a date can be moved up and down, dated events are always in
> the
> same order. After applying changes by user new order is sent to the server
> where new order is being validated and commited.
> 
> Page for updating a person is slightly changed. Event positing in the list
> (a.k.a. weight) is exposed to the form. Form hadling code uses this weight
> to
> reconstitute order of the events and to distinguish old events from new
> ones.
> Before that all new events were automatically sorted. A warning about
> changed
> event order is removed.
> 
> ## Also type alias for cdate was removed.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/geneanet/geneweb/pull/349
> Commit Summary
> - Fix issue 336. Reordering events
> 
> File Changes
> - _A_ hd/etc/eventsorder.txt
>   https://github.com/geneanet/geneweb/pull/349/files#diff-0 (98)
> - _M_ hd/etc/perso_module/timeline.txt
>   https://github.com/geneanet/geneweb/pull/349/files#diff-1 (4)
> - _M_ hd/etc/updind.txt
>   https://github.com/geneanet/geneweb/pull/349/files#diff-2 (2)
> - _M_ src/.depend
>   https://github.com/geneanet/geneweb/pull/349/files#diff-3 (4)
> - _M_ src/adef.mli
>   https://github.com/geneanet/geneweb/pull/349/files#diff-4 (5)
> - _M_ src/check.ml
>   https://github.com/geneanet/geneweb/pull/349/files#diff-5 (6)
> - _M_ src/checkItem.ml
>   https://github.com/geneanet/geneweb/pull/349/files#diff-6 (5)
> - _M_ src/checkItem.mli
>   https://github.com/geneanet/geneweb/pull/349/files#diff-7 (2)
> - _M_ src/def.mli
>   https://github.com/geneanet/geneweb/pull/349/files#diff-8 (5)
> - _M_ src/hutil.mli
>   https://github.com/geneanet/geneweb/pull/349/files#diff-9 (1)
> - _M_ src/mergeIndOk.ml
>   https://github.com/geneanet/geneweb/pull/349/files#diff-10 (6)
> - _M_ src/mutil.ml
>   https://github.com/geneanet/geneweb/pull/349/files#diff-11 (7)
> - _M_ src/mutil.mli
>   https://github.com/geneanet/geneweb/pull/349/files#diff-12 (1)
> - _M_ src/perso.ml
>   https://github.com/geneanet/geneweb/pull/349/files#diff-13 (35)
> - _M_ src/perso.mli
>   https://github.com/geneanet/geneweb/pull/349/files#diff-14 (1)
> - _M_ src/update.ml
>   https://github.com/geneanet/geneweb/pull/349/files#diff-15 (2)
> - _M_ src/updateInd.ml
>   https://github.com/geneanet/geneweb/pull/349/files#diff-16 (43)
> - _M_ src/updateInd.mli
>   https://github.com/geneanet/geneweb/pull/349/files#diff-17 (5)
> - _M_ src/updateIndOk.ml
>   https://github.com/geneanet/geneweb/pull/349/files#diff-18 (205)
> - _M_ src/updateIndOk.mli
>   https://github.com/geneanet/geneweb/pull/349/files#diff-19 (3)
> - _M_ src/util.ml
>   https://github.com/geneanet/geneweb/pull/349/files#diff-20 (17)
> - _M_ src/util.mli
>   https://github.com/geneanet/geneweb/pull/349/files#diff-21 (7)
> 
> Patch Links:
> - https://github.com/geneanet/geneweb/pull/349.patch
> - https://github.com/geneanet/geneweb/pull/349.diff
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/geneanet/geneweb/pull/349.

Kakadu commented:
@A2line , english, please

Kakadu committed:
Fix issue 336. Reordering events

Timeline template has now a link to the dedicated page for events reordering.
Events without a date (or text date) can be moved up and down, dated events
are always in the same order (except events with exactly same date: they can
be swapped only with each other). After applying changes by user new order
is sent to the server where new order is being validated and commited.

Page for updating a person is slightly changed. Event positing in the list
(a.k.a. weight) is exposed to the form. Form hadling code uses this weight to
reconstitute order of the events and to distinguish old events from new ones.
Before that all new events were automatically sorted. A warning about changed
event order is removed.

Also type alias for cdate was removed.

a2line commented:
My bad.

I add a event with defaut template and when I try to open new “Edit person
events order” link, I always have an empty answer from the server that says
“Incorrect request”.

2015-08-27 16:26 GMT+02:00 Kakadu notifications@github.com:

> @A2line https://github.com/A2line , english, please
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/geneanet/geneweb/pull/349#issuecomment-135451994.

swdes commented:
Hello @Kakadu 
It seems OK with the basic test db grimaldi.ged
You should just add event/fact description in the timeline editor screen.
![image](https://cloud.githubusercontent.com/assets/5917598/9584514/bf81475c-5011-11e5-9422-8c5f516ae21a.png)
I'll make further test tommorrow.

a2line commented:
It's ok for me now (when I make distrib kakadu:event-order)

Ordering is not allowed with custom events or custom dates and that are the cases that will need reordering the most. It looks like death event isn't reorderer automatically before burial, see:
![events-order](https://cloud.githubusercontent.com/assets/10353895/9611358/8354d8d8-50df-11e5-8f4a-823c6c6bd4e7.JPG)

Kakadu commented:
> It looks like death event isn't reorderer automatically before burial, see:

Yes, it is not and it is not expected to be reordered. [See](https://github.com/geneanet/geneweb/issues/336#issuecomment-132249340)

> undated cremation or burial always goes at the end

Well, I think I can treat event with text date as undated one to allow reordering.

Can you, please, briefly explain how to create precise event? I was trying to set `Exact` and year 1900 but some reason this information is erased when changes are applied. I think I can't guess how to use it without looking at code. Should precise events be automatically sorted like events with normal date? I don't know yet but intuition says that these dates are only partially orderable.

Also, when we create custom event we need to select some type  and than select 1st item to make visible text box at the right side. Is it the same for you?

a2line commented:
Right. I confirm custom event without date can be orderdered, and with date are ordered well. It's only missing when we add custom dates.

Yes, like you with default template, I need to make it in two steps to have the text box after “-” to create a custom event; under templm it's a bit better managed. It's counterintuitive, we need to show the field text if no event is selected in perso.txt for default template.

a2line commented:
(mistake in previous post: s/perso.txt/updind.txt/)

I was testing further with last templm for gw7 that manage events easier and found another new bug link to events (dunno if it's linked to this pull req or another recent commit):
– when we edit an individual (MOD_IND) the formulary always add the Birth event and it looks like it waits for a date (it shouldn't!). If no date is given, gwd gives error “**Assertion failed, file updateIndOk.ml, chars 1463-11**” ; 
– same when we add a Death event, if we don't add date, the event is not stored so we can't indicate that someone is dead without precising date. It's really anoying!

Kakadu commented:
I found some weird issue when we remove all events from MOD_IND page. I will look at it tomorrow.

a2line commented:
Ok for custom dates. I think now it would be nice to be able to handle
family events ordering (“Edit family events order”). We have to merge
family events with personnal ones. At the moment except the marriage, all
events add themselves after the list of personnals events. When we have
multiple families it's quickly a mess in event orders on personnal
timeline.

An easy way to manage this could be to split personnal timeline from a new
family timeline, but do we want to have two timelines?

2015-09-02 0:33 GMT+02:00 Kakadu notifications@github.com:

> I found some weird issue when we remove all events from MOD_IND page. I
> will look at it tomorrow.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/geneanet/geneweb/pull/349#issuecomment-136882352.

swdes commented:
@A2line Thanks to have discoverd this issue. I think it could become to complex to mix both event ordering... We have to think carefully before any implementation.

swdes commented:
We missed a point. When 2 events have the same date, you should be able to order them. How do you think we could  improve your PR to cover this need ?
![image](https://cloud.githubusercontent.com/assets/5917598/9720522/26ed398c-558f-11e5-9462-e58ea9026f29.png)

Kakadu commented:
Well, I can easily allow reordering on server side but in this case user will be able to broke date-based event sorting. It seems that right way is to add some checks on client side too. I will try something...

Kakadu commented:
https://github.com/kakadu/geneweb/tree/reorder-same-dates

@swdes, I think it is what you need.

swdes commented:
@Kakadu, perfect for us

swdes commented:
@kakadu, in order to merge your pull request, could you:
- merge "reorder same dates" in your "event-order" branch
- clean the git history (with interactive rebase) in order to have one or two significant commits
  Thanks

Kakadu commented:
@swdes , rebased.

swdes commented:
Thanks, I've just merged it

a2line commented:
The case I talked about of a death event without any other datas is not
solved.

Test this : edit family, add a child with just first name, edit this new
individual, choose “Died…” in Death subsection, validate. Nothing is stored.

Worse, if you edit an existing individual already stored as dead, then
validate, we loose the data (the field goes back to “-”). This behavior is
wrong. We were able to use this before and I have a lot of individuals from
oral sources on whom I only know the name plus the fact that they are dead
which is a data in itself. I don't want to loose it by inadvertance because
Geneweb is tweaking the field without my knowledge.

swdes commented:
You're absolutely right.
@Kakadu could you fix it ?

a2line commented:
I'm trying to add an event (custom or not) with a custom date and when I
wan't to order it, I see “Move up”/“Move down” but if I click them, nothing
move.

PS : don't hate me Kakadu :p, I love what you do! Thank you for your work,
we miss people like you able to understand Ocaml and tweak Geneweb.

Kakadu commented:
I have fixed death disappearance in https://github.com/Kakadu/geneweb/tree/event-order. This commit is not in the pull request because (I think) this one is already closed.

Also, I don't fully understand this death type combo box, for example, 3rd option has three dots and it seems that it should unlock some additional controls (or alll other options should hide some controls). `Dead as an infant` option is very likely should not allow user to enter any concrete date of death. 5th option is most funny: internal text says `OfCourseDead` but text says `may be dead`.

Kakadu commented:
Reordering events with  custom text date is fixed (same branch)

a2line commented:
Both commits are good. Thank you.

From what I remember, templm uses 2nd and 5th death options looking at the
age of the individual. For exemple if between 80 and 120 years old, instead
of “-” it automatically goes to “Unknown” and above 120  “Certainly dead”.
Both are only shown in ind update formulary (and not on the individual
page), so it's only a kind of note for wizards.
